python/tank/platform/software_launcher.py

```diff
@@ -239,38 +276,156 @@ def minimum_supported_version(self):
+
+        Example::
+
+            ["15.5.324", "16.0.1.322"]
+        """
+        return self._versions
+
+    ##########################################################################################
+    # abstract methods
 
     def prepare_launch(self, exec_path, args, file_to_open=None):
         """
         Prepares the given software for launch
 
         :param str exec_path: Path to DCC executable to launch
         :param str args: Command line arguments as strings
-        :param str file_to_open: (optional) Full path name of a file to open on launch
+        :param str file_to_open: (optional) Full path name of a file to open on
+            launch
         :returns: :class:`LaunchInformation` instance
         """
         raise NotImplementedError
 
+    def _scan_software(self):
+        """
+        Abstract method that must be implemented by subclasses to return a list
+        of ``SoftwareVersion`` objects on the local filesystem.
+
+        Typical implementations will use functionality such as ``glob`` to
+        locate all versions and variations of executables on disk and then
+        creating ``SoftwareVersion`` objects for each executable. This method
+        is called by the ``get_supported_software()`` method, and each
+        ``SoftwareVersion`` object returned is checked against the launcher's
+        lists of supported version and product variations via the
+        ``_is_supported`` method.
+        """
+
+        raise NotImplementedError
+
+    def _scan_software_with_expression(self, match_template, regular_expression):
+        """
+        This is a helper method that can be invoked in an implementation of :meth:`_scan_software`.
+
+        The ``match_template`` argument provides a template to use both for globing files and then pattern
+        matching them using regular expressions provided by the ``tokens`` dictionary.
+
+        The method will first substitute every token from the template with a ``*`` for globing files.
+        It will then replace the tokens in the template with the regular expressions that were
+        provided.
+
+        In the following example::
+
+            self._scan_software_with_expression(
+                "C:/Program Files/Nuke{full_version}/Nuke{major_minor_version}.exe",
+                {
+                    "full_version": r"(?P<version>[\d.v]+)",
+                    "major_minor_version": r"(?P<major_minor_version>[\d.]+)"
+                }
+            )
+
+        this would first look for every files matching C:\Program Files\Nuke*\Nuke*.exe and then
+        run the regular expression C:\\Program Files\\Nuke(?P<version>[\d.v]+)\\(?P<major_minor_version>[\d.]+).
+        and would return any files matching that pattern as well as any extracted tokens.
+
+        For the method to extract an argument from the matched string and return it back to the caller
+        the ``?P<variable>`` notation must be used when invoking this method.
```

Comment:
I don't think TDs are familiar with this notation (i don't know it very well!). What value does it add?  I think we should return a list instead in order to keep things a bit simpler.